pkg/apiserver/certificate/cacert_controller_test.go

```diff
@@ -0,0 +1,566 @@
+			expectedErrMsg: "error getting ValidatingWebhookConfiguration crdvalidator.antrea.io: validatingwebhookconfigurations.admissionregistration.k8s.io \"crdvalidator.antrea.io\" not found",
+		},
+		{
+			name: "fail to update webhook",
+			existingWebhook: &admissionregistrationv1.ValidatingWebhookConfiguration{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "crdvalidator.antrea.io",
+				},
+				Webhooks: []admissionregistrationv1.ValidatingWebhook{
+					{
+						Name: "acnpmutator.antrea.io",
+					},
+				},
+			},
+			prepareReactor: func(clientset *fakeclientset.Clientset) {
+				clientset.PrependReactor("update", "validatingwebhookconfigurations", func(action cgtesting.Action) (handled bool, ret runtime.Object, err error) {
+					return true, &admissionregistrationv1.ValidatingWebhookConfiguration{}, errors.New("error updating validatingwebhookconfigurations")
+				})
+			},
+			expectedErrMsg: "error updating antrea CA cert of ValidatingWebhookConfiguration crdvalidator.antrea.io: error updating validatingwebhookconfigurations",
+		},
+	}
+	caConfig := &CAConfig{
+		ServiceName:     AntreaServiceName,
+		PairName:        "antrea-controller",
+		CAConfigMapName: "antrea-ca",
+		ValidatingWebhooks: []string{
+			"crdvalidator.antrea.io",
+		},
+	}
+	secureServing := genericoptions.NewSecureServingOptions().WithLoopback()
+	var err error
+	caConfig.SelfSignedCertDir, err = os.MkdirTemp("", "antrea-self-signed")
+	if err != nil {
+		t.Fatalf("Unable to create temporary directory: %v", err)
+	}
+	defer os.RemoveAll(caConfig.SelfSignedCertDir)
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			clientset := fakeclientset.NewSimpleClientset()
+			aggregatorClientset := fakeaggregatorclientset.NewSimpleClientset()
+			apiExtensionClient := fakeapiextensionclientset.NewSimpleClientset()
+			caContentProvider, _ := generateSelfSignedCertificate(secureServing, caConfig)
+
+			if tt.existingWebhook != nil {
+				clientset = fakeclientset.NewSimpleClientset(tt.existingWebhook)
+			}
+			tt.prepareReactor(clientset)
+			controller := newCACertController(caContentProvider, clientset, aggregatorClientset, apiExtensionClient, caConfig)
+			caBundle := []byte("abc")
+			err = controller.syncValidatingWebhooks(caBundle)
+			if tt.expectedErrMsg != "" {
+				require.ErrorContains(t, err, tt.expectedErrMsg)
+				return
+			}
+			require.NoError(t, err)
+			vWebhook, err := clientset.AdmissionregistrationV1().ValidatingWebhookConfigurations().Get(context.Background(), "crdvalidator.antrea.io", metav1.GetOptions{})
+			require.NoError(t, err)
+			for _, webhook := range vWebhook.Webhooks {
+				assert.Equal(t, caBundle, webhook.ClientConfig.CABundle)
+			}
+		})
+	}
+}
+
+func TestSyncMutatingWebhooks(t *testing.T) {
+	tests := []struct {
+		name                    string
+		existingWebhook         *admissionregistrationv1.MutatingWebhookConfiguration
+		existingOptionalWebhook *admissionregistrationv1.MutatingWebhookConfiguration
```

Comment:
Is it simpler to just use an existingWebhook slice, which represents the objects that will be added to storage before testing?

Reply:
Yeah, refined, thanks.

Reply:
didn't see update

Reply:
Ah, misunderstood the comments, merge them into a slice. thanks.